images/executor/src/code-caller.ts

```diff
@@ -0,0 +1,668 @@
+          String(error)
+      );
+    }
+    this._checkState();
+    debug(
+      `exit _handleChildError(), state: ${String(this.state)}, uuid: ${
+        this.uuid
+      }`
+    );
+  }
+
+  _timeout() {
+    debug(`enter _timeout(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+    this._checkState([CallerState.IN_CALL]);
+    this.timeoutID = null;
+    const err = new Error("timeout exceeded, killing PythonCaller child");
+    this.child.kill("SIGTERM");
+    this.state = CallerState.EXITING;
+    this._callCallback(err);
+    debug(`exit _timeout(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+  }
+
+  _clearTimeout() {
+    debug(
+      `enter _clearTimeout(), state: ${String(this.state)}, uuid: ${this.uuid}`
+    );
+    clearTimeout(this.timeoutID);
+    this.timeoutID = null;
+    debug(
+      `exit _clearTimeout(), state: ${String(this.state)}, uuid: ${this.uuid}`
+    );
+  }
+
+  _callCallback(err: Error, data?: any, output?: string) {
+    debug(
+      `enter _callCallback(), state: ${String(this.state)}, uuid: ${this.uuid}`
+    );
+    if (err) data = this._errorData();
+    const c = this.callback;
+    this.callback = null;
+    c(err, data, output);
+    debug(
+      `exit _callCallback(), state: ${String(this.state)}, uuid: ${this.uuid}`
+    );
+  }
+
+  _callIsFinished() {
+    debug(
+      `enter _callIsFinished(), state: ${String(this.state)}, uuid: ${
+        this.uuid
+      }`
+    );
+    if (!this._checkState([CallerState.IN_CALL])) return;
+    this._clearTimeout();
+    let data,
+      err = null;
+    try {
+      data = JSON.parse(this.outputData);
+    } catch (e) {
+      err = new Error("Error decoding PythonCaller JSON: " + e.message);
+    }
+    if (err) {
+      this.state = CallerState.EXITING;
+      this.child.kill("SIGTERM");
+      this._callCallback(err);
+    } else {
+      this.state = CallerState.WAITING;
+      if (data.present) {
+        this._callCallback(null, data.val, this.outputBoth);
+      } else {
+        this._callCallback(
+          new FunctionMissingError("Function not found in module")
+        );
+      }
+    }
+    debug(
+      `exit _callIsFinished(), state: ${String(this.state)}, uuid: ${this.uuid}`
+    );
+  }
+
+  _errorData() {
+    const errForStack = new Error();
+    return {
+      state: this.state,
+      childIsNull: this.child == null,
+      callbackIsNull: this.callback == null,
+      timeoutIDIsNull: this.timeoutID == null,
+      outputStdout: this.outputStdout,
+      outputStderr: this.outputStderr,
+      outputBoth: this.outputBoth,
+      outputData: this.outputData,
+      stack: errForStack.stack,
+      lastCallData: this.lastCallData
+    };
+  }
+
+  _logError(msg: string) {
+    debug(
+      `enter _logError(), state: ${String(this.state)}, uuid: ${this.uuid}`
+    );
+    const errData = this._errorData();
+    console.error(msg, errData);
```

Comment:
Will this write JSON to STDERR? If not, won’t this result in an “Error decoding JSON” from `code-caller-docker`?